File: godot-loader/src/nativeMain/kotlin/godot/loader/internal/NativeKObject.kt
package godot.loader.internal

import jni.JObject
import jni.JniEnv
import kotlinx.cinterop.COpaquePointer
import kotlinx.cinterop.toCPointer

class NativeKObject(_wrapped: JObject) {
    val wrapped = _wrapped.newGlobalRef()

    fun _onInit(env: JniEnv) {
        val initMethod = getMethodId("_onInit", "()V")
        wrapped.callVoidMethod(initMethod)
    }

    fun _onDestroy(env: JniEnv) {
        val destroyMethod = getMethodId("_onDestroy", "()V")
        wrapped.callVoidMethod(destroyMethod)
        dispose()
    }

    fun getRawPtr(env: JniEnv): COpaquePointer {
        val getClassNameMethod = getMethodId("_getRawPtr", "()J")
        return wrapped.callLongMethod(getClassNameMethod).toCPointer()!!
    }

    fun dispose() {
        wrapped.deleteGlobalRef()
    }

    @ThreadLocal
    companion object : JObjectWrapper("godot.internal.KObject")
}

File: godot-loader2/src/native_kobject.cpp
#include "native_kobject.h"

JClassHelper NativeKObject::JH = JClassHelper("godot.internal.KObject");

void NativeKObject::init(jni::Env& env, jni::JObject object) {
    // so it won't be gc'd by jvm
    wrapped = object.newGlobalRef(env);
}

void NativeKObject::dispose(jni::Env& env) {
    wrapped.deleteGlobalRef(env);
}

void NativeKObject::_onInit(jni::Env& env, jni::JObject classLoader) {

}

void NativeKObject::_onDestroy(jni::Env& env, jni::JObject classLoader) {

}
